### PR TITLE
dm-master/: add offline master operation, combine offline-worker into offline-member

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
https://github.com/pingcap/dm/issues/653

### What is changed and how it works?
1. add `offline-member` command in dmctl.
2. add offline master operation

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes

 - Need to update the `dm/dm-ansible`
 - Need to be included in the release note

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@lichunzhu please resolve conflicts

File: dm/master/server_test.go [hunk 1]
```diff
@@ -994,47 +994,138 @@ func (t *testMaster) TestOperateSource(c *check.C) {
...
+	c.Assert(listResp.Members[1].Name, check.Equals, cfg2.Name)
+
+	_, leaderID2, _, err := s1.election.LeaderInfo(ctx)
+	c.Assert(err, check.IsNil)
+
+	if leaderID == cfg3.Name {
+		// s3 is leader, leader should re-campaign
+		c.Assert(leaderID != leaderID2, check.IsTrue)
+	} else {
+		// s3 is leader, leader should keep the same
```
Comment (csuzhangxc) [inline dm/master/server_test.go:1101]: `s3 is leader`?

File: dm/ctl/master/offline_member.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
+// Copyright 2019 PingCAP, Inc.
```
[the remark this answers is not in the document]
Comment (lichunzhu) [reply dm/ctl/master/offline_member.go:1]: oops

File: dm/ctl/master/offline_member.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	"github.com/pingcap/dm/dm/pb"
+
+	"github.com/pingcap/errors"
+	"github.com/spf13/cobra"
+)
+
+// NewOfflineMemberCmd creates an OfflineWorker command
+func NewOfflineMemberCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "offline-member [--master/--worker] [--name master-name/worker-name]",
```
[the remark this answers is not in the document]
Comment (lichunzhu) [reply dm/ctl/master/offline_member.go:30]: Yes. I will change this to <>

File: dm/ctl/master/offline_member.go [hunk 4]
```diff
@@ -27,7 +27,7 @@ import (
 // NewOfflineMemberCmd creates an OfflineWorker command
 func NewOfflineMemberCmd() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:   "offline-member [--master/--worker] [--name master-name/worker-name]",
+		Use:   "offline-member <--master/--worker> [--name master-name/worker-name]",
```
Comment (csuzhangxc) [inline dm/ctl/master/offline_member.go:30]: `--name` should also be `<>`, right?

**Comment by csuzhangxc** [comment]
I ran `make generate_mock`, got 😂 
```diff
Index: dm/pbmock/dmmaster.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- dm/pbmock/dmmaster.go	(revision f872b5a78e6f7df0ad26c02433193e707473cde2)
+++ dm/pbmock/dmmaster.go	(date 1590026813182)
@@ -6,11 +6,10 @@
 
 import (
 	context "context"
-	reflect "reflect"
-
 	gomock "github.com/golang/mock/gomock"
 	pb "github.com/pingcap/dm/dm/pb"
 	grpc "google.golang.org/grpc"
+	reflect "reflect"
 )
 
 // MockMasterClient is a mock of MasterClient interface
```

**Comment by lichunzhu** [comment]
> I ran `make generate_mock`, got 😂
> 
> ```diff
> Index: dm/pbmock/dmmaster.go
> IDEA additional info:
> Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
> <+>UTF-8
> ===================================================================
> --- dm/pbmock/dmmaster.go	(revision f872b5a78e6f7df0ad26c02433193e707473cde2)
> +++ dm/pbmock/dmmaster.go	(date 1590026813182)
> @@ -6,11 +6,10 @@
>  
>  import (
>  	context "context"
> -	reflect "reflect"
> -
>  	gomock "github.com/golang/mock/gomock"
>  	pb "github.com/pingcap/dm/dm/pb"
>  	grpc "google.golang.org/grpc"
> +	reflect "reflect"
>  )
>  
>  // MockMasterClient is a mock of MasterClient interface
> ```

goimports changes this. I will fix it.

File: dm/ctl/master/offline_member.go (hunk 4 above)
Comment (lichunzhu) [reply dm/ctl/master/offline_member.go:30]: Yes

**Comment by csuzhangxc** [comment]
when I ran unit tests, I got
```
FAIL: election_test.go:213: testElectionSuite.TestElection2After1

election_test.go:215:
    testElection2After1(t, c, false)
election_test.go:166:
    // for abnormally exited election, session will be cleared here
    c.Assert(deleted, IsTrue)
... obtained bool = false
```

**Comment by lichunzhu** [comment]
> when I ran unit tests, I got
> 
> ```
> FAIL: election_test.go:213: testElectionSuite.TestElection2After1
> 
> election_test.go:215:
>     testElection2After1(t, c, false)
> election_test.go:166:
>     // for abnormally exited election, session will be cleared here
>     c.Assert(deleted, IsTrue)
> ... obtained bool = false
> ```

That's weired.. Can you represent some logs? I can't reproduce this on my machine.

File: dm/ctl/master/offline_member.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+		Run:   offlineMemberFunc,
+	}
+	cmd.Flags().BoolP("master", "m", false, "to offline a master")
+	cmd.Flags().BoolP("worker", "w", false, "to offline a worker")
+	cmd.Flags().StringP("name", "n", "", "specify member name for choosing type")
+	return cmd
+}
+
+func convertOfflineMemberType(cmd *cobra.Command) (string, error) {
+	master, err := cmd.Flags().GetBool("master")
```
Comment (WangXiangUSTC) [inline dm/ctl/master/offline_member.go:41]: how about defining const variables for "master" and "worker"?

File: tests/_utils/test_prepare [hunk 6]
```diff
@@ -100,3 +100,12 @@ function dmctl_operate_source() {
         "\"result\": true" 2 \
         "\"source\": \"$source_id\"" 1
 }
+
+
+function dmctl_offline_member() {
```
Comment (WangXiangUSTC) [inline tests/_utils/test_prepare:105]: seems that this function is not used

File: pkg/election/election.go [hunk 7]
```diff
@@ -385,13 +391,42 @@ forLoop:
...
+		if leaderInfo.ID == id {
+			deleteKey = string(kv.Key)
+			break
+		}
+	}
+	if deleteKey == "" {
+		// no campaign info left in etcd, no need to trigger re-campaign
+		return false, nil
+	}
+	delResp, err := e.cli.Delete(ctx, deleteKey)
```
Comment (WangXiangUSTC) [inline pkg/election/election.go:416]: what error message will get for `elec.Campaign`(L244) when deleting the key?

Comment (lichunzhu) [reply pkg/election/election.go:416]: It will get no error but a delete event

File: tests/_utils/test_prepare (hunk 6 above)
Comment (lichunzhu) [reply tests/_utils/test_prepare]: deleted

File: dm/ctl/master/offline_member.go (hunk 5 above)
Comment (lichunzhu) [reply dm/ctl/master/offline_member.go:41]: I think we shouldn't use const variables for flags. But the type is using consts now

**Pull request merged**